src/util/ieee_float.cpp

```diff
@@ -0,0 +1,167 @@
+/*
+ * Conversion of float to IEEE-754 and vice versa.
+ *
+ * © Copyright 2018 Pedro Gimeno Fortea.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the
+ * "Software"), to deal in the Software without restriction, including
+ * without limitation the rights to use, copy, modify, merge, publish,
+ * distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject to
+ * the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+#include "log.h"
+#include "porting.h"
+#include <stdint.h>
+#include <math.h>
+
+// Given an unsigned 32-bit integer representing an IEEE-754 single-precision
+// float, return the float.
+float u32_FloatSlow(uint32_t i)
+{
+	float res;
+	int exp = (i >> 23) & 0xFF;
+	uint32_t sign = i & 0x80000000UL;
+	uint32_t imant = i & 0x7FFFFFUL;
+	if (exp == 0xFF) {
+		// Inf/NaN
+		if (imant == 0)
+			return sign ? -1.f/0.f : 1.f/0.f;
+		return 0.f/0.f; // always return indeterminate
+	}
+
+	if (!exp) {
+		// Denormal or zero
+		return sign ? -ldexpf((float)imant, -149) : ldexpf((float)imant, -149);
+	}
+
+	return sign ? -ldexpf((float)(imant | 0x800000), exp - 150) :
+		ldexpf((float)(imant | 0x800000), exp - 150);
```

Comment:
These two 0x800000 should have been 0x800000UL. I missed them, sorry. It shouldn't matter much, though, because they should be converted to unsigned anyway, and it's safe to assume that ints are > 16 bits.